File: backend/internal/dto/dto.go
package dto

// OpenAI Chat Completions request/response (subset)

type OpenAIChatRequest struct {
        Model            string            `json:"model"`
        Messages         []OpenAIMessage   `json:"messages"`
        MaxTokens        *int              `json:"max_tokens,omitempty"`
        Temperature      *float64          `json:"temperature,omitempty"`
        TopP             *float64          `json:"top_p,omitempty"`
        Stream           bool              `json:"stream,omitempty"`
        Stop             any               `json:"stop,omitempty"`
        Tools            []OpenAITool      `json:"tools,omitempty"`
        ToolChoice       any               `json:"tool_choice,omitempty"`
        User             string            `json:"user,omitempty"`
        FrequencyPenalty *float64          `json:"frequency_penalty,omitempty"`
        PresencePenalty  *float64          `json:"presence_penalty,omitempty"`
        N                *int              `json:"n,omitempty"`
        ResponseFormat   any               `json:"response_format,omitempty"`
        StreamOptions    *StreamOptions    `json:"stream_options,omitempty"`
}

type StreamOptions struct {
        IncludeUsage bool `json:"include_usage,omitempty"`
}

type OpenAIMessage struct {
        Role       string `json:"role"`
        Content    any    `json:"content"` // string or []content parts
        Name       string `json:"name,omitempty"`
        ToolCalls  []OpenAIToolCall `json:"tool_calls,omitempty"`
        ToolCallID string `json:"tool_call_id,omitempty"`
}

type OpenAITool struct {
        Type     string         `json:"type"`
        Function OpenAIFunction `json:"function"`
}

type OpenAIFunction struct {
        Name        string `json:"name"`
        Description string `json:"description,omitempty"`
        Parameters  any    `json:"parameters,omitempty"`
}

type OpenAIToolCall struct {
        ID       string `json:"id"`
        Type     string `json:"type"`
        Function struct {
                Name      string `json:"name"`
                Arguments string `json:"arguments"`
        } `json:"function"`
        Index *int `json:"index,omitempty"`
}

type OpenAIChatResponse struct {
        ID      string         `json:"id"`
        Object  string         `json:"object"`
        Created int64          `json:"created"`
        Model   string         `json:"model"`
        Choices []OpenAIChoice `json:"choices"`
        Usage   *OpenAIUsage   `json:"usage,omitempty"`
}

type OpenAIChoice struct {
        Index        int           `json:"index"`
        Message      OpenAIMessage `json:"message,omitempty"`
        Delta        OpenAIMessage `json:"delta,omitempty"`
        FinishReason *string       `json:"finish_reason"`
}

type OpenAIUsage struct {
        PromptTokens        int                  `json:"prompt_tokens"`
        CompletionTokens    int                  `json:"completion_tokens"`
        TotalTokens         int                  `json:"total_tokens"`
        PromptTokensDetails *PromptTokensDetails `json:"prompt_tokens_details,omitempty"`
}

type PromptTokensDetails struct {
        CachedTokens int `json:"cached_tokens,omitempty"`
}

// Claude Messages API (subset)

type ClaudeRequest struct {
        Model         string          `json:"model"`
        Messages      []ClaudeMessage `json:"messages"`
        MaxTokens     int             `json:"max_tokens"`
        System        any             `json:"system,omitempty"` // string or []blocks
        Temperature   *float64        `json:"temperature,omitempty"`
        TopP          *float64        `json:"top_p,omitempty"`
        Stream        bool            `json:"stream,omitempty"`
        StopSequences []string        `json:"stop_sequences,omitempty"`
        Tools         []ClaudeTool    `json:"tools,omitempty"`
        ToolChoice    any             `json:"tool_choice,omitempty"`
}

type ClaudeMessage struct {
        Role    string `json:"role"`
        Content any    `json:"content"` // string or []ClaudeContentBlock
}

type ClaudeContentBlock struct {
        Type      string          `json:"type"`
        Text      string          `json:"text,omitempty"`
        ID        string          `json:"id,omitempty"`
        Name      string          `json:"name,omitempty"`
        Input     any             `json:"input,omitempty"`
        ToolUseID string          `json:"tool_use_id,omitempty"`
        Content   any             `json:"content,omitempty"`
        Source    *ClaudeImageSrc `json:"source,omitempty"`
}

type ClaudeImageSrc struct {
        Type      string `json:"type"`
        MediaType string `json:"media_type,omitempty"`
        Data      string `json:"data,omitempty"`
}

type ClaudeTool struct {
        Name        string `json:"name"`
        Description string `json:"description,omitempty"`
        InputSchema any    `json:"input_schema,omitempty"`
}

type ClaudeResponse struct {
        ID           string               `json:"id"`
        Type         string               `json:"type"`
        Role         string               `json:"role"`
        Content      []ClaudeContentBlock `json:"content"`
        Model        string               `json:"model"`
        StopReason   string               `json:"stop_reason"`
        StopSequence *string              `json:"stop_sequence"`
        Usage        ClaudeUsage          `json:"usage"`
}

type ClaudeUsage struct {
        InputTokens              int `json:"input_tokens"`
        OutputTokens             int `json:"output_tokens"`
        CacheReadInputTokens     int `json:"cache_read_input_tokens,omitempty"`
        CacheCreationInputTokens int `json:"cache_creation_input_tokens,omitempty"`
}

// Models list (OpenAI format)
type ModelsListResponse struct {
        Object string       `json:"object"`
        Data   []ModelItem  `json:"data"`
}

type ModelItem struct {
        ID      string `json:"id"`
        Object  string `json:"object"`
        Created int64  `json:"created"`
        OwnedBy string `json:"owned_by"`
}

File: backend/src/dto.rs
//! OpenAI / Claude wire formats (subset), mirroring Go internal/dto.
//! Unknown fields are ignored on input; output field presence matches the
//! Go structs' omitempty behavior.

use serde::{Deserialize, Serialize};
use serde_json::Value;

fn is_false(b: &bool) -> bool {
    !*b
}
fn is_zero(n: &i64) -> bool {
    *n == 0
}
#[allow(clippy::ptr_arg)] // serde skip_serializing_if passes &String
fn is_empty_str(s: &String) -> bool {
    s.is_empty()
}

// ---- OpenAI ----

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAIChatRequest {
    pub model: String,
    pub messages: Vec<OpenAIMessage>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_p: Option<f64>,
    #[serde(skip_serializing_if = "is_false")]
    pub stream: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop: Option<Value>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub tools: Vec<OpenAITool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_choice: Option<Value>,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub user: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frequency_penalty: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub presence_penalty: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub n: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_format: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream_options: Option<StreamOptions>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct StreamOptions {
    #[serde(skip_serializing_if = "is_false")]
    pub include_usage: bool,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAIMessage {
    pub role: String,
    /// string or []content parts; always serialized (null when absent), like Go.
    pub content: Option<Value>,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub name: String,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub tool_calls: Vec<OpenAIToolCall>,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub tool_call_id: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAITool {
    #[serde(rename = "type")]
    pub tool_type: String,
    pub function: OpenAIFunction,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAIFunction {
    pub name: String,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parameters: Option<Value>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAIToolCall {
    pub id: String,
    #[serde(rename = "type")]
    pub call_type: String,
    pub function: OpenAIToolCallFunction,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub index: Option<i64>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAIToolCallFunction {
    pub name: String,
    pub arguments: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAIChatResponse {
    pub id: String,
    pub object: String,
    pub created: i64,
    pub model: String,
    pub choices: Vec<OpenAIChoice>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub usage: Option<OpenAIUsage>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAIChoice {
    pub index: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<OpenAIMessage>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub delta: Option<OpenAIMessage>,
    /// Always serialized (null when absent), like Go.
    pub finish_reason: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct OpenAIUsage {
    pub prompt_tokens: i64,
    pub completion_tokens: i64,
    pub total_tokens: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_tokens_details: Option<PromptTokensDetails>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct PromptTokensDetails {
    #[serde(skip_serializing_if = "is_zero")]
    pub cached_tokens: i64,
}

// ---- Claude ----

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct ClaudeRequest {
    pub model: String,
    pub messages: Vec<ClaudeMessage>,
    pub max_tokens: i64,
    /// string or []blocks
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_p: Option<f64>,
    #[serde(skip_serializing_if = "is_false")]
    pub stream: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub stop_sequences: Vec<String>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub tools: Vec<ClaudeTool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_choice: Option<Value>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct ClaudeMessage {
    pub role: String,
    /// string or []ClaudeContentBlock
    pub content: Option<Value>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct ClaudeContentBlock {
    #[serde(rename = "type")]
    pub block_type: String,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub text: String,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub id: String,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input: Option<Value>,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub tool_use_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source: Option<ClaudeImageSrc>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct ClaudeImageSrc {
    #[serde(rename = "type")]
    pub src_type: String,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub media_type: String,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub data: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct ClaudeTool {
    pub name: String,
    #[serde(skip_serializing_if = "is_empty_str")]
    pub description: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_schema: Option<Value>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct ClaudeResponse {
    pub id: String,
    #[serde(rename = "type")]
    pub resp_type: String,
    pub role: String,
    pub content: Vec<ClaudeContentBlock>,
    pub model: String,
    /// Go emits "" when empty; upstream may send null.
    #[serde(deserialize_with = "de_null_string")]
    pub stop_reason: String,
    /// Always serialized (null when absent), like Go.
    pub stop_sequence: Option<String>,
    pub usage: ClaudeUsage,
}

fn de_null_string<'de, D: serde::Deserializer<'de>>(d: D) -> Result<String, D::Error> {
    Ok(Option::<String>::deserialize(d)?.unwrap_or_default())
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(default)]
pub struct ClaudeUsage {
    pub input_tokens: i64,
    pub output_tokens: i64,
    #[serde(skip_serializing_if = "is_zero")]
    pub cache_read_input_tokens: i64,
    #[serde(skip_serializing_if = "is_zero")]
    pub cache_creation_input_tokens: i64,
}

// ---- Models list (OpenAI format) ----

#[derive(Debug, Clone, Serialize)]
pub struct ModelsListResponse {
    pub object: String,
    pub data: Vec<ModelItem>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ModelItem {
    pub id: String,
    pub object: String,
    pub created: i64,
    pub owned_by: String,
}
